lib/AST/ParameterPack.cpp

```diff
@@ -314,15 +314,36 @@ CanPackType PackType::getReducedShape() {
 }
 
 CanType TypeBase::getReducedShape() {
-  if (auto *packArchetype = getAs<PackArchetypeType>())
-    return packArchetype->getReducedShape();
-
   if (auto *packType = getAs<PackType>())
     return packType->getReducedShape();
 
   if (auto *expansionType = getAs<PackExpansionType>())
     return expansionType->getReducedShape();
 
+  struct PatternTypeWalker : public TypeWalker {
+    CanType shapeType;
+
+    Action walkToTypePre(Type type) override {
```

Comment:
The PatternTypeWalker looks like it's the same thing as calling TypeBase::getTypeParameterPacks() and then getting the reduced shape of one of the results? However now I see a bug in TypeBase::getTypeParameterPacks():

```
      if (archetypeTy->isRoot())
        typeParams.insert(paramTy);
```

if your pattern only references a non-root archetype, its going to skip it. We should include non-root archetypes in the result too. This is called from resolvePackExpansionType() in TypeCheckType.cpp among other places, which means we missed test cases involving associated types of type parameter packs

Reply:
Ah, you're right, thanks! I totally forgot about `getTypeParameterPacks`. I'll fix the issue with pack references via associated types